File: py/2020/day23/part2.py
import numpy as np

inp = "315679824"
cups_n = 1000000


def move_cups(cups):
    p, c1, c2, c3 = cups[:4]
    cups_ = cups[4:]

    label = 0
    for i in range(1, cups_n):
        label = (p-i) % (cups_n + 1)
        if label not in [p, c1, c2, c3, 0]:
            break
    index = np.where(cups_==label)[0][0]
    xs = cups_[:(index+1)]
    ys = cups_[(index+1):]

    return np.concatenate((xs, np.array([c1, c2, c3]), ys, np.array([p])), axis=None)

cups = list(map(int, list(inp))) + list(range(10, cups_n+1))
cups = np.array(cups)

moves = 10000000

for _ in range(moves):
    cups = move_cups(cups)

j = np.where(cups==1)[0][0]

print(cups[j+1] * cups[j+2])

File: hs/2020/day23/part2.hs
import AOC
import Data.Char
import Data.List
import Data.Maybe
import Data.Vector (Vector)
import qualified Data.Vector as Vector

cupNumber :: Int
cupNumber = 1000000

modulo :: Integral a => a -> a -> a
modulo a b = ((mod a b) + b) `mod` b

moveCups :: Vector Int -> Vector Int
moveCups cups = Vector.concat [xs, Vector.fromList threegroup, ys, Vector.fromList [p]]
    where p = Vector.head cups
          c1 = cups Vector.! 1
          c2 = cups Vector.! 2
          c3 = cups Vector.! 3
          threegroup = [c1, c2, c3]
          cups' = Vector.drop 4 cups
          labels = map ((`modulo` (cupNumber + 1)) . (p-)) [1..cupNumber]
          index = case head (filter (not . null) (map (`Vector.elemIndex` cups') labels)) of Just n -> n
          (xs, ys) = Vector.splitAt (index + 1) cups'

main = do
    let inp = map digitToInt "315679824" :: [Int]
    let cups = inp ++ [10..cupNumber]
    let finalCups = iterate moveCups (Vector.fromList cups) !! 1000
    let index = Vector.head $ 1 `Vector.elemIndices` finalCups
    let a = finalCups Vector.! ((index+1) `mod` cupNumber)
    let b = finalCups Vector.! ((index+2) `mod` cupNumber)
    print a
    print b
